### PR TITLE
Spectator Camera V2

### PR DESCRIPTION
**Pull request by zfox23**
This PR adds the following:
- Users can now capture 360 snapshots from the Spectator Camera app.
- Users can now capture still snapshots from the Spectator Camera's point-of-view (i.e. selfies).
- Users can now zoom in and out while using the Spectator Camera.
- The Spectator Camera app is now an official, certified .app.json Marketplace app (however, it won't be uploaded to the Marketplace until the release containing this code is out).
- The Spectator Camera app is now usable in both VR and Desktop modes.
- The Snapshot API has been updated to allow for scripts and developers to take 360 snapshots from a given camera location in either equirectangular or cubemap format.

If you want to use the `Spectator Camera v2.0` app before it's out on the Marketplace:
    1. Download [this folder](https://www.dropbox.com/sh/udlx0afoknqmd9e/AABY_M1DdaKsoLTXwiRZCmDxa?dl=0) from Dropbox as a ZIP, then extract its contents anywhere on your computer.
    2. Drag `spectatorCamera.js` from <unzipped "v2.0" directory> onto Interface from this PR
    3. Say "Yes" when asked whether or not you want to run the script

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9661/app_13141_PR_0236bc0c4484c818266cb7ac720da072614e585e.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9661/HighFidelity-Beta-PR13141-0236bc0c4484c818266cb7ac720da072614e585e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9661/HighFidelity-Beta-Interface-PR13141-0236bc0c4484c818266cb7ac720da072614e585e.exe

File: interface/src/ui/Snapshot.h [hunk 1]
```diff
@@ -51,11 +54,28 @@ class Snapshot : public QObject, public Dependency {
...
+    static qint16 snapshotIndex;
+    static bool oldEnabled;
+    static QVariant oldAttachedEntityId;
+    static QVariant oldOrientation;
+    static QVariant oldvFoV;
+    static QVariant oldNearClipPlaneDistance;
+    static QVariant oldFarClipPlaneDistance;
+    static QImage imageArray[6];
+    static void convertToCubemap();
+    static void convertToEquirectangular();
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.h]: Why are those all statics? There is only a single`Snapshot` object, those should probably all be standard members.

File: interface/src/ui/Snapshot.cpp [hunk 2]
```diff
@@ -46,6 +54,14 @@ const QString URL = "highfidelity_url";
 
 Setting::Handle<QString> Snapshot::snapshotsLocation("snapshotsLocation");
 
+QTimer Snapshot::snapshotTimer;
+Snapshot::Snapshot() {
+    Snapshot::snapshotTimer.setSingleShot(false);
+    Snapshot::snapshotTimer.setTimerType(Qt::PreciseTimer);
+    Snapshot::snapshotTimer.setInterval(250);
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.cpp]: Coding Standard: Magic number

File: interface/src/ui/Snapshot.cpp [hunk 3]
```diff
@@ -88,6 +104,230 @@ QString Snapshot::saveSnapshot(QImage image, const QString& filename, const QStr
...
+
+    // Initialize some secondary camera render config options for 360 snapshot capture
+    static_cast<ToneMappingConfig*>(qApp->getRenderEngine()->getConfiguration()->getConfig("SecondaryCameraJob.ToneMapping"))->setCurve(0);
+
+    secondaryCameraRenderConfig->resetSizeSpectatorCamera(static_cast<int>(CUBEMAP_SIDE_PIXEL_DIMENSION), static_cast<int>(CUBEMAP_SIDE_PIXEL_DIMENSION));
+    secondaryCameraRenderConfig->setProperty("attachedEntityId", "");
+    secondaryCameraRenderConfig->setPosition(cameraPosition);
+    secondaryCameraRenderConfig->setProperty("vFoV", 90.0f);
+    secondaryCameraRenderConfig->setProperty("nearClipPlaneDistance", 0.3f);
+    secondaryCameraRenderConfig->setProperty("farClipPlaneDistance", 16384.0f);
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.cpp]: Coding Standard: magic numbers

File: interface/src/ui/Snapshot.cpp [hunk 4]
```diff
@@ -88,6 +104,230 @@ QString Snapshot::saveSnapshot(QImage image, const QString& filename, const QStr
...
+
+    // Setup for Down Image capture
+    secondaryCameraRenderConfig->setOrientation(glm::quat(glm::radians(glm::vec3(-90.0f, 0.0f, 0.0f))));
+
+    Snapshot::snapshotIndex = 0;
+
+    Snapshot::snapshotTimer.start(350);
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.cpp]: Coding Standard: magic number

File: interface/src/ui/Snapshot.cpp [hunk 5]
```diff
@@ -88,6 +104,230 @@ QString Snapshot::saveSnapshot(QImage image, const QString& filename, const QStr
...
+
+    secondaryCameraRenderConfig->resetSizeSpectatorCamera(static_cast<int>(CUBEMAP_SIDE_PIXEL_DIMENSION), static_cast<int>(CUBEMAP_SIDE_PIXEL_DIMENSION));
+    secondaryCameraRenderConfig->setProperty("attachedEntityId", "");
+    secondaryCameraRenderConfig->setPosition(cameraPosition);
+    secondaryCameraRenderConfig->setProperty("vFoV", 90.0f);
+    secondaryCameraRenderConfig->setProperty("nearClipPlaneDistance", 0.3f);
+    secondaryCameraRenderConfig->setProperty("farClipPlaneDistance", 16384.0f);
+
+    // Setup for Down Image capture
+    secondaryCameraRenderConfig->setOrientation(glm::quat(glm::radians(glm::vec3(-90.0f, 0.0f, 0.0f))));
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.cpp]: Coding Standard: magic number, it'd be nice to name that quat `Y_DOWN` or something like that to make it clearer.

File: interface/src/ui/Snapshot.cpp [hunk 6]
```diff
@@ -88,6 +104,230 @@ QString Snapshot::saveSnapshot(QImage image, const QString& filename, const QStr
...
+            z = glm::cos(phi) * glm::sin(theta) * -1.0f;
+
+            float xa, ya, za;
+            float a;
+
+            a = std::max(std::max(std::abs(x), std::abs(y)), std::abs(z));
+
+            xa = x / a;
+            ya = y / a;
+            za = z / a;
```
Comment (Atlante45) [inline interface/src/ui/Snapshot.cpp]: I understand this comes from the SO post, but it'd be nice to not unnecessarily split all those declarations/definitions.
It just opens the doors to potential maintenance errors.

File: interface/src/Application.cpp [hunk 7]
```diff
@@ -7494,6 +7494,12 @@ void Application::takeSecondaryCameraSnapshot(const QString& filename) {
     });
 }
 
+void Application::takeSecondaryCamera360Snapshot(const glm::vec3& cameraPosition, const bool& cubemapOutputFormat, const QString& filename) {
+    postLambdaEvent([filename, cubemapOutputFormat, cameraPosition, this] {
```
Comment (Atlante45) [inline interface/src/Application.cpp]: `this` seems unnecessary here.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9777/HighFidelity-Beta-PR13141-0c766593244bb253ead3f6e8fcc6bca6f80bfa2e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9777/HighFidelity-Beta-PR13141-0c766593244bb253ead3f6e8fcc6bca6f80bfa2e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9777/HighFidelity-Beta-Interface-PR13141-0c766593244bb253ead3f6e8fcc6bca6f80bfa2e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9777/HighFidelity-Beta-PR13141-0c766593244bb253ead3f6e8fcc6bca6f80bfa2e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9777/HighFidelity-Beta-Interface-PR13141-0c766593244bb253ead3f6e8fcc6bca6f80bfa2e.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9782/HighFidelity-Beta-PR13141-e607c68bea14074276267b6a2563764fd4da25e9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9782/HighFidelity-Beta-PR13141-e607c68bea14074276267b6a2563764fd4da25e9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9782/HighFidelity-Beta-Interface-PR13141-e607c68bea14074276267b6a2563764fd4da25e9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9782/HighFidelity-Beta-PR13141-e607c68bea14074276267b6a2563764fd4da25e9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9782/HighFidelity-Beta-Interface-PR13141-e607c68bea14074276267b6a2563764fd4da25e9.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9786/HighFidelity-Beta-PR13141-a54ce96ac2a1e35ad02473471a3174e815f8b9a3.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9786/HighFidelity-Beta-PR13141-a54ce96ac2a1e35ad02473471a3174e815f8b9a3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9786/HighFidelity-Beta-Interface-PR13141-a54ce96ac2a1e35ad02473471a3174e815f8b9a3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9786/HighFidelity-Beta-PR13141-a54ce96ac2a1e35ad02473471a3174e815f8b9a3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9786/HighFidelity-Beta-Interface-PR13141-a54ce96ac2a1e35ad02473471a3174e815f8b9a3.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9799/HighFidelity-Beta-PR13141-906678123ca1659976736c532b21add62e009885.apk

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9799/HighFidelity-Beta-Interface-PR13141-906678123ca1659976736c532b21add62e009885.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9799/HighFidelity-Beta-PR13141-906678123ca1659976736c532b21add62e009885.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9799/HighFidelity-Beta-Interface-PR13141-906678123ca1659976736c532b21add62e009885.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9809/HighFidelity-Beta-PR13141-bccaa2dbe57d727b316dd9aa01f8f69a5ab42483.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9809/HighFidelity-Beta-Interface-PR13141-bccaa2dbe57d727b316dd9aa01f8f69a5ab42483.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9809/HighFidelity-Beta-PR13141-bccaa2dbe57d727b316dd9aa01f8f69a5ab42483.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9809/HighFidelity-Beta-Interface-PR13141-bccaa2dbe57d727b316dd9aa01f8f69a5ab42483.exe

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9817/HighFidelity-Beta-Interface-PR13141-96324984226d0e78d28d72c382cfe06fe2138fbe.dmg

**Comment by zfox23** [comment]
## Test Plan
### Prerequisites
- Start the Test Plan in Desktop mode.
- You'll be asked to wear an HMD at some points during the test, so have an HMD ready.
- Have the `Spectator Camera v2` app running:
    1. Download [this folder](https://www.dropbox.com/sh/udlx0afoknqmd9e/AABY_M1DdaKsoLTXwiRZCmDxa?dl=0) from Dropbox as a ZIP, then extract its contents anywhere on your computer.
    2. Drag `spectatorCamera.js` from <unzipped "v2.0" directory> onto Interface from this PR
    3. Say "Yes" when asked whether or not you want to run the script
- Install your favorite 360 photo viewer (I like [Insta360 Player](https://static.insta360.com/software/Player/Windows/Insta360Player_Win_v2.3.6.exe))
- In `Settings -> General`, ensure the "Put my snapshots here" text field is blank.
- You're in an interesting domain that you're excited to take snapshots in 😄 

### Tests
#### Smoke Test
Before performing any of the "official" tests below, please try out the Spectator app on your own, both in HMD mode and in Desktop mode. See if you can figure out how to use it without any test plan instructions, and mark down any anomalies or difficulties you run into (if any).

Note that the app's interface and usage is different in VR mode vs Desktop mode.

#### Spectator Camera v2.0 Test
1. Verify that you see a "SPECTATOR" app icon in your HUD.
2. Click the "SPECTATOR" app icon in your HUD.
    i. You see the Spectator Camera app.
    ii. The on/off master switch in the top right of the app is in the "off" position (switch is on the left and its background is white)
    iii. You see a staticy preview, two disabled "SNAP" buttons, an app description, and a "Change Snapshot Location" link.
3. Click the "Learn More about Spectator" link towards the bottom of the app.
    i. You see a white popup explaining more about the app.
4. Click the "OBS Official Overview Guide" link.
    i. Your default browser has opened outside Interface and is showing you an OBS guide.
5. Click outside the white popup.
    i. The popup is closed.
6. Click the "Change Snapshot Location" link.
    i. You are taken to the General Settings page, where you see a "Put my snapshots here" settings field
7. Turn on the Spectator Camera using the switch at the top right of the app.
    i. The switch turns green.
    ii. A camera model shows up in front of you in-world. The camera model shows a preview of what it is seeing in its "viewfinder".
    iii. You see a preview of the camera's view from within the app.
    iv. The two "SNAP" buttons are enabled.
    v. You see a "Field of View" slider below the camera preview.
8. Attempt to move the camera model with the mouse.
    i. As you move the camera model, the camera preview from within the app changes to reflect its new position.
9. Click the "SNAP PICTURE" button.
    i. You hear a snapshot sound.
    ii. You see a window that prompts you to choose your snapshot location.
10. Choose a snapshot location (i.e. your Desktop) and click "CHOOSE". Use Windows Explorer to navigate to that folder.
    i. Verify that you see a `.jpg` snapshot taken from the point of view of the camera in that directory.
11. Click the "SNAP 360" button.
    i. You hear a snapshot sound.
    ii. The app says "PROCESSING" for a couple of seconds, and you can't click on any buttons in the app.
    iii. After a couple of seconds, you see a `.jpg` 360 snapshot taken from the location of the camera in your snapshots directory
12. Open your 360 photo viewer (i.e. Insta360 Player). Drag that `.jpg` 360 snapshot into the viewer. Scroll around the image.
    i. Verify that the 360 snapshot is super cool and that it looks correct.
13. Back in Interface, drag the Field of View slider left and right.
    i. Verify that the camera's view zooms in and out.
    ii. Verify that the Field of View number (in degrees) updates in real-time.
14. Click the "reset" button to the right of the Field of View slider.
    i. The Field of View resets to 45 degrees.
15. Put on your HMD.
    i. The Spectator Camera model remains in-world.
16. Open your Tablet and find the SPECTATOR app.
    i. Verify that the app's icon is "active" (i.e. green).
17. Click the SPECTATOR app icon in your Tablet.
    i. The switch in the top right is "on" (green).
    ii. The Spectator Camera model remains in-world. The camera model shows a preview of what it is seeing in its "viewfinder".
    iii. You see a preview of the camera's view from within the app.
    iv. The two "SNAP" buttons are enabled.
    v. You see a "Field of View" slider below the camera preview.
    vi. You see a "Monitor Shows" radio button field at the top of the camera preview, and "Camera View" is selected.
    vii. You see a "Shortcuts" area in the app. There are two checkboxes under "Shortcuts", and they are both UNCHECKED:
        - "Left Thumbstick"/"Left Thumbpad" (depending on your HMD)
        - "Right Thumbstick"/"Right Thumbpad" (depending on your HMD)
18. Using your left hand controller, click in the left thumbstick/thumbpad.
    i. Nothing happens (if you're on a Vive, you might see the "teleport" overlay appear briefly).
19. Using your right hand controller, click in the right thumbstick/thumbpad.
    i. Nothing happens (if you're on a Vive, you might see the "teleport" overlay appear briefly).
20. Check both checkboxes under "Shortcuts", then close and re-open the Spectator app.
    i. The two checkboxes under "Shortcuts" are still CHECKED.
21. Using your left hand controller, click in the left thumbstick/thumbpad.
    i. The "Monitor Shows" radio button field changed - now "VR Preview" is the checked option.
    ii. Peek under your HMD - verify that what you see on your monitor in indeed a preview of what you're looking at in VR.
22. Using your left hand controller, click in the left thumbstick/thumbpad again.
    i. The "Monitor Shows" radio button field changed - now "Camera View" is the checked option.
    ii. Peek under your HMD - verify that what you see on your monitor in indeed a preview of what the Spectator Camera is "seeing".
23. Using your left hand, grab the Camera and maneuver it such that it's pointed at you (i.e. a selfie!). Say "CHEESE!" Using your right hand controller, click in the right thumbstick/thumbpad.
    i. You hear a snapshot sound.
    ii. (You can check this later if that's more convenient) You see a `.jpg` snapshot taken from the camera's point of view in your snapshots directory.
24. Let go of the camera. Click the "SNAP 360" button.
    i. You hear a snapshot sound.
    ii. The app says "PROCESSING" for a couple of seconds, and you can't click on any buttons in the app.
    iii. (You can check this later if that's more convenient) After a couple of seconds, you see a `.jpg` 360 snapshot taken from the location of the camera in your snapshots directory.

#### JS API Tests
1. Turn off the Spectator Camera. Write down the XYZ coordinates of an "interesting location" in your domain (for example, `(0, 0, 0)`, or your avatar's current location (discoverable from GOTO)). Open the JS console (`CTRL + ALT + J`). Input the following command in the JS console and press enter (replacing what's in the angle brackets below):
    i. `Window.takeSecondaryCamera360Snapshot(JSON.stringify({x: <x coordinate of interesting location>, y: <y coordinate of interesting location>, z: <z coordinate of interesting location>}))`
2. Wait a few moments.
    i. Verify that you see a notification in the top right of your screen that a snapshot has been saved.
    ii. You see a `.jpg` 360 snapshot taken from the XYZ location from (1) in your snapshots directory
    iii. The 360 snapshot is in equirectangular format (i.e. its width is double its height)
3. Input the following command in the JS console and press enter (replacing what's in the angle brackets below):
    i. `Window.takeSecondaryCamera360Snapshot(JSON.stringify({x: <x coordinate of interesting location>, y: <y coordinate of interesting location>, z: <z coordinate of interesting location>}), true)`
4. Wait a few moments.
    i. Verify that you see a notification in the top right of your screen that a snapshot has been saved.
    ii. You see a `.jpg` 360 snapshot taken from the XYZ location from (1) in your snapshots directory.
    iii. The 360 snapshot is in cubemap format (i.e. you see a cube's six "faces" in the image and its height is 75% of its width).

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9821/HighFidelity-Beta-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9821/HighFidelity-Beta-Interface-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.dmg

**Comment by zfox23** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9836/HighFidelity-Beta-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9836/HighFidelity-Beta-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9836/HighFidelity-Beta-Interface-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9836/HighFidelity-Beta-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9836/HighFidelity-Beta-Interface-PR13141-d904bcd1cd8011ecc56cc1241e1cfa7df9ac47a3.exe

**Comment by qlikmorales** [comment]
Working on Desktop/HMD

**Comment by zfox23** [comment]
From offline discussion - not a fail. Please continue testing.

**Comment by qlikmorales** [comment]
Passed @conklin94122

**Pull request merged**